### **Paso 5.2.7 - Ingesta del directorio "laptimes" para el directorio "2021-03-21"**

Nos permite crear e indicar parámetros en tiempo de ejecución

<center><img src="https://i.postimg.cc/Sx9fDwyH/db151.png"></center>

In [ ]:
dbutils.widgets.text("p_data_source", "")
v_data_source = dbutils.widgets.get("p_data_source")

In [ ]:
v_data_source

Out[3]: 'Eargast'

In [ ]:
dbutils.widgets.text("p_file_date", "2021-03-21")
v_file_date = dbutils.widgets.get("p_file_date")

In [ ]:
v_file_date

Out[5]: '2021-03-21'

In [ ]:
%run "../includes/configuration"

In [ ]:
%run "../includes/common_functions"

#### Paso 1 - Leer el directorio **lap_times** el cual contiene multiples archivos CSV

In [ ]:
from pyspark.sql.types import StructType, StructField, IntegerType, StringType

In [ ]:
lap_times_schema = StructType(fields=[StructField("raceId", IntegerType(), False),
                                      StructField("driverId", IntegerType(), True),
                                      StructField("lap", IntegerType(), True),
                                      StructField("position", IntegerType(), True),
                                      StructField("time", StringType(), True),
                                      StructField("milliseconds", IntegerType(), True)
                                     ])

In [ ]:
# El parámetro "raw_folder_path" se encuentra en el notebook "configuration"
# El parámetro "v_file_date" se encuentra en el notebook e indicamos su valor en tiempo de ejecución
lap_times_df = spark.read \
.schema(lap_times_schema) \
.csv(f"{raw_folder_path}/{v_file_date}/lap_times")
# Esto seria equivalente a la ruta: /mnt/formula1dl/raw/2021-03-21/laptimes

In [ ]:
lap_times_df.show(truncate=False)

+------+--------+---+--------+--------+------------+
|raceId|driverId|lap|position|time    |milliseconds|
+------+--------+---+--------+--------+------------+
|841   |20      |1  |1       |1:38.109|98109       |
|841   |20      |2  |1       |1:33.006|93006       |
|841   |20      |3  |1       |1:32.713|92713       |
|841   |20      |4  |1       |1:32.803|92803       |
|841   |20      |5  |1       |1:32.342|92342       |
|841   |20      |6  |1       |1:32.605|92605       |
|841   |20      |7  |1       |1:32.502|92502       |
|841   |20      |8  |1       |1:32.537|92537       |
|841   |20      |9  |1       |1:33.240|93240       |
|841   |20      |10 |1       |1:32.572|92572       |
|841   |20      |11 |1       |1:32.669|92669       |
|841   |20      |12 |1       |1:32.902|92902       |
|841   |20      |13 |1       |1:33.698|93698       |
|841   |20      |14 |3       |1:52.075|112075      |
|841   |20      |15 |4       |1:38.385|98385       |
|841   |20      |16 |2       |1:31.548|91548  

In [ ]:
lap_times_df.printSchema()

root
 |-- raceId: integer (nullable = true)
 |-- driverId: integer (nullable = true)
 |-- lap: integer (nullable = true)
 |-- position: integer (nullable = true)
 |-- time: string (nullable = true)
 |-- milliseconds: integer (nullable = true)



#### Paso 2 - Renombrar columnas y añadir nuevas columnas
1. Renombrar driverId y raceId
2. Añadir ingestion_date con current timestamp

In [ ]:
from pyspark.sql.functions import lit

In [ ]:
# La función "add_ingestion_date()" se encuentra en el notebook "common_functions"
lap_times_with_ingestion_date_df = add_ingestion_date(lap_times_df)

In [ ]:
lap_times_with_ingestion_date_df.show(truncate=False)

+------+--------+---+--------+--------+------------+-----------------------+
|raceId|driverId|lap|position|time    |milliseconds|ingestion_date         |
+------+--------+---+--------+--------+------------+-----------------------+
|841   |20      |1  |1       |1:38.109|98109       |2023-06-15 17:28:01.211|
|841   |20      |2  |1       |1:33.006|93006       |2023-06-15 17:28:01.211|
|841   |20      |3  |1       |1:32.713|92713       |2023-06-15 17:28:01.211|
|841   |20      |4  |1       |1:32.803|92803       |2023-06-15 17:28:01.211|
|841   |20      |5  |1       |1:32.342|92342       |2023-06-15 17:28:01.211|
|841   |20      |6  |1       |1:32.605|92605       |2023-06-15 17:28:01.211|
|841   |20      |7  |1       |1:32.502|92502       |2023-06-15 17:28:01.211|
|841   |20      |8  |1       |1:32.537|92537       |2023-06-15 17:28:01.211|
|841   |20      |9  |1       |1:33.240|93240       |2023-06-15 17:28:01.211|
|841   |20      |10 |1       |1:32.572|92572       |2023-06-15 17:28:01.211|

In [ ]:
final_df = lap_times_with_ingestion_date_df.withColumnRenamed("driverId", "driver_id") \
                                           .withColumnRenamed("raceId", "race_id") \
                                           .withColumn("data_source", lit(v_data_source)) \
                                           .withColumn("file_date", lit(v_file_date))

In [ ]:
final_df.show(truncate=False)

+-------+---------+---+--------+--------+------------+-----------------------+-----------+----------+
|race_id|driver_id|lap|position|time    |milliseconds|ingestion_date         |data_source|file_date |
+-------+---------+---+--------+--------+------------+-----------------------+-----------+----------+
|841    |20       |1  |1       |1:38.109|98109       |2023-06-15 17:28:01.793|Eargast    |2021-03-21|
|841    |20       |2  |1       |1:33.006|93006       |2023-06-15 17:28:01.793|Eargast    |2021-03-21|
|841    |20       |3  |1       |1:32.713|92713       |2023-06-15 17:28:01.793|Eargast    |2021-03-21|
|841    |20       |4  |1       |1:32.803|92803       |2023-06-15 17:28:01.793|Eargast    |2021-03-21|
|841    |20       |5  |1       |1:32.342|92342       |2023-06-15 17:28:01.793|Eargast    |2021-03-21|
|841    |20       |6  |1       |1:32.605|92605       |2023-06-15 17:28:01.793|Eargast    |2021-03-21|
|841    |20       |7  |1       |1:32.502|92502       |2023-06-15 17:28:01.793|Earg

#### Paso 3 - Escribir datos en el datalake como parquet y crear la tabla **lap_times** en la base de datos **f1_processed**

In [ ]:
final_df.select('race_id').distinct().collect()

Out[18]: [Row(race_id=858),
 Row(race_id=897),
 Row(race_id=879),
 Row(race_id=883),
 Row(race_id=898),
 Row(race_id=853),
 Row(race_id=918),
 Row(race_id=857),
 Row(race_id=876),
 Row(race_id=847),
 Row(race_id=874),
 Row(race_id=842),
 Row(race_id=914),
 Row(race_id=844),
 Row(race_id=908),
 Row(race_id=916),
 Row(race_id=860),
 Row(race_id=926),
 Row(race_id=875),
 Row(race_id=855),
 Row(race_id=861),
 Row(race_id=882),
 Row(race_id=906),
 Row(race_id=912),
 Row(race_id=896),
 Row(race_id=852),
 Row(race_id=887),
 Row(race_id=867),
 Row(race_id=881),
 Row(race_id=868),
 Row(race_id=846),
 Row(race_id=871),
 Row(race_id=886),
 Row(race_id=904),
 Row(race_id=931),
 Row(race_id=845),
 Row(race_id=929),
 Row(race_id=849),
 Row(race_id=843),
 Row(race_id=854),
 Row(race_id=862),
 Row(race_id=903),
 Row(race_id=891),
 Row(race_id=850),
 Row(race_id=880),
 Row(race_id=870),
 Row(race_id=900),
 Row(race_id=884),
 Row(race_id=890),
 Row(race_id=930),
 Row(race_id=877),
 Row(race_id=856),
 Ro

In [ ]:
for race_id_list in final_df.select('race_id').distinct().collect():
  print(race_id_list)

Row(race_id=858)
Row(race_id=897)
Row(race_id=879)
Row(race_id=883)
Row(race_id=898)
Row(race_id=853)
Row(race_id=918)
Row(race_id=857)
Row(race_id=876)
Row(race_id=847)
Row(race_id=874)
Row(race_id=842)
Row(race_id=914)
Row(race_id=844)
Row(race_id=908)
Row(race_id=916)
Row(race_id=860)
Row(race_id=926)
Row(race_id=875)
Row(race_id=855)
Row(race_id=861)
Row(race_id=882)
Row(race_id=906)
Row(race_id=912)
Row(race_id=896)
Row(race_id=852)
Row(race_id=887)
Row(race_id=867)
Row(race_id=881)
Row(race_id=868)
Row(race_id=846)
Row(race_id=871)
Row(race_id=886)
Row(race_id=904)
Row(race_id=931)
Row(race_id=845)
Row(race_id=929)
Row(race_id=849)
Row(race_id=843)
Row(race_id=854)
Row(race_id=862)
Row(race_id=903)
Row(race_id=891)
Row(race_id=850)
Row(race_id=880)
Row(race_id=870)
Row(race_id=900)
Row(race_id=884)
Row(race_id=890)
Row(race_id=930)
Row(race_id=877)
Row(race_id=856)
Row(race_id=928)
Row(race_id=869)
Row(race_id=885)
Row(race_id=907)
Row(race_id=848)
Row(race_id=895)
Row(race_id=89

In [ ]:
for race_id_list in final_df.select('race_id').distinct().collect():
  print(race_id_list.race_id)

858
897
879
883
898
853
918
857
876
847
874
842
914
844
908
916
860
926
875
855
861
882
906
912
896
852
887
867
881
868
846
871
886
904
931
845
929
849
843
854
862
903
891
850
880
870
900
884
890
930
877
856
928
869
885
907
848
895
893
866
911
933
863
859
899
865
913
872
888
910
902
909
892
905
901
915
873
894
864
851
927
932
841
878
211
193
939
210
183
192
159
236
223
222
209
230
225
232
233
190
224
177
185
206
212
182
218
205
168
178
142
164
169
944
235
191
163
227
936
165
207
179
189
937
197
202
231
175
196
948
217
229
220
173
161
176
162
238
171
194
166
234
239
938
216
181
945
167
237
160
219
143
226
214
941
195
221
203
141
943
200
145
940
228
213
170
188
204
198
199
158
201
942
184
186
215
174
172
144
949
934
180
208
187
31
961
950
34
28
26
27
44
12
350
22
346
47
353
355
955
1
52
13
348
6
16
3
40
20
340
339
48
5
959
19
41
347
43
15
37
9
17
35
960
4
8
39
23
49
7
958
343
956
51
957
69
10
50
951
45
952
38
25
24
70
351
29
21
338
352
345
32
341
953
354
33
11
68
14
344
349
342
42
337
2


In [ ]:
n = 0
for race_id_list in final_df.select('race_id').distinct().collect():
  n = n + 1
print(n)

454


1. Si no hubiesemos utilizado la sentencia **IF** y la tabla **f1_processed.lap_times** no hubiese existido, nos hubiese devuelto un **ERROR**. Es por eso que utilizamos el **IF**
2. **spark.conf.set("spark.sql.sources.partitionOverwriteMode","dynamic")** nos permite utilizar la función **insertInto** en modo **overwrite** y asi SOLO SOBRESCRIBIR las particiones que sean coincidentes. Tengo un ejemplo en mi material de **Apache Spark con Python**
3. Por lo tanto, si la tabla ya existe, solo sobrescribirá las particiones coincidentes
4. Y si no existe la tabla, la creará
5. Como estamos ingestando el directorio de archivos CSV **lap_times** del directorio **2021-03-21**, existen 454 particiones
6. Con este método, podemos EJECUTAR EL NOTEBOOK CUANTAS VECES QUERAMOS Y NO SE DUPLICARAN LOS DATOS

In [ ]:
spark.conf.set("spark.sql.sources.partitionOverwriteMode","dynamic")

In [ ]:
final_df = final_df.select("driver_id","lap","position","time","milliseconds","ingestion_date","data_source","file_date","race_id")

In [ ]:
if (spark._jsparkSession.catalog().tableExists("f1_processed.lap_times")):
  final_df.write.mode('overwrite').insertInto("f1_processed.lap_times")
else:
  final_df.write.partitionBy('race_id').format('parquet').saveAsTable("f1_processed.lap_times")

In [ ]:
spark.read.parquet("/mnt/formula1dl/processed/lap_times").show(truncate=False)

+---------+---+--------+--------+------------+-----------------------+-----------+----------+-------+
|driver_id|lap|position|time    |milliseconds|ingestion_date         |data_source|file_date |race_id|
+---------+---+--------+--------+------------+-----------------------+-----------+----------+-------+
|4        |1  |1       |1:22.551|82551       |2023-06-15 17:28:14.555|Eargast    |2021-03-21|59     |
|4        |2  |1       |1:18.209|78209       |2023-06-15 17:28:14.555|Eargast    |2021-03-21|59     |
|4        |3  |1       |1:17.071|77071       |2023-06-15 17:28:14.555|Eargast    |2021-03-21|59     |
|4        |4  |1       |1:16.751|76751       |2023-06-15 17:28:14.555|Eargast    |2021-03-21|59     |
|4        |5  |1       |1:16.472|76472       |2023-06-15 17:28:14.555|Eargast    |2021-03-21|59     |
|4        |6  |1       |1:16.826|76826       |2023-06-15 17:28:14.555|Eargast    |2021-03-21|59     |
|4        |7  |1       |1:16.698|76698       |2023-06-15 17:28:14.555|Eargast    |

In [ ]:
%sql
SELECT * FROM f1_processed.lap_times;

driver_id lap position time milliseconds ingestion_date data_source file_date race_id 4 1 1 1:22.551 82551 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 2 1 1:18.209 78209 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 3 1 1:17.071 77071 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 4 1 1:16.751 76751 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 5 1 1:16.472 76472 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 6 1 1:16.826 76826 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 7 1 1:16.698 76698 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 8 1 1:16.854 76854 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 9 1 1:16.734 76734 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 10 1 1:15.924 75924 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 11 1 1:15.671 75671 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 12 1 1:16.318 76318 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 13 1 1:17.243 77243 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 14 1 1:16.605 76605 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 15 1 1:17.602 77602 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 16 1 1:16.534 76534 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 17 1 1:15.906 75906 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 18 1 1:15.680 75680 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 19 1 1:15.763 75763 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 20 1 1:16.405 76405 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 21 1 1:16.951 76951 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 22 1 1:16.231 76231 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 23 1 1:16.048 76048 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 24 2 1:34.559 94559 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 25 1 1:19.613 79613 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 26 1 1:17.980 77980 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 27 1 1:17.825 77825 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 28 1 1:17.914 77914 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 29 1 1:17.838 77838 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 30 1 1:18.348 78348 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 31 1 1:18.226 78226 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 32 1 1:18.399 78399 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 33 1 1:18.798 78798 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 34 1 1:18.489 78489 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 35 1 1:16.883 76883 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 36 1 1:17.555 77555 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 37 1 1:17.812 77812 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 38 1 1:19.705 79705 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 39 1 1:17.132 77132 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 40 1 1:19.843 79843 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 41 1 1:19.045 79045 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 42 1 1:17.815 77815 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 43 1 1:17.436 77436 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 44 1 1:16.385 76385 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 45 1 1:16.403 76403 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 46 1 1:16.999 76999 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 47 1 1:17.727 77727 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 48 1 1:16.626 76626 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 49 1 1:35.310 95310 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 50 1 2:16.067 136067 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 51 1 2:01.666 121666 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 52 1 2:00.278 120278 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 53 1 1:17.759 77759 2023-06-15T17:28:14.555+0000 Eargast 2021-03-21 59 4 54 1 1:17.631 77631 2023-06-15T17:28:14.555+0000 

In [ ]:
%sql
-- Vemos que tenemos la data hasta la race_id = 1047
SELECT race_id, COUNT(1)
FROM f1_processed.lap_times
GROUP BY race_id
ORDER BY race_id DESC

race_id count(1) 1047 1043 1046 1531 1045 1016 1044 1076 1043 1128 1042 1288 1041 1017 1040 946 1039 778 1038 924 1037 766 1036 1274 1035 1025 1034 895 1033 1327 1032 1226 1031 1140 1030 1075 1029 1381 1028 1030 1027 1370 1026 986 1025 900 1024 1162 1023 990 1022 784 1021 1358 1020 1054 1019 913 1018 1401 1017 1036 1016 1310 1015 1489 1014 1274 1013 947 1012 1046 1011 1081 1010 1038 1009 938 1008 1318 1007 1242 1006 931 1005 976 1004 948 1003 1145 1002 925 1001 690 1000 1230 999 1250 998 898 997 1239 996 915 995 1182 994 1513 993 1016 992 841 991 1115 990 997 989 937 988 1031 987 1167 986 1225 985 933 984 884 983 1024 982 845 981 1001 980 780 979 1255 978 889 977 1197 976 824 975 1172 974 1414 973 1108 972 828 971 885 970 903 969 942 968 1008 967 1292 966 1479 965 1078 964 1155 963 1040 962 1167 961 1020 960 756 959 1416 958 1511 957 970 956 1416 955 1028 954 1371 953 1357 952 1234 951 973 950 1223 949 1002 948 1017 945 1033 944 1327 943 1261 942 837 941 848 940 1042 939 1068 938 935 937 736 936 1307 934 721 933 1062 932 1331 931 1429 930 1242 929 1019 928 1012 927 942 926 703 918 1012 917 1231 916 852 915 1030 914 909 913 1141 912 1084 911 858 910 1226 909 1310 908 920 907 1453 906 1174 905 1280 904 1361 903 1140 902 1113 901 980 900 947 899 1441 898 1170 897 1145 896 1201 895 1006 894 1150 893 1268 892 1069 891 890 890 1394 888 1182 887 1109 886 1460 885 1503 884 1323 883 1206 882 1049 881 1089 880 1124 879 1395 878 1250 877 1083 876 1368 875 1126 874 1088 873 1280 872 1189 871 820 870 1621 869 1580 868 1097 867 1210 866 1537 865 1446 864 1359 863 1289 862 1291 861 1270 860 1044 859 1552 858 1172 857 1186 856 1205 855 1221 854 1335 853 889 852 888 851 1489 850 1273 849 1075 848 1343 847 1488 846 1604 845 1491 844 1302 843 1284 842 1128 841 1083 355 1187 354 1649 353 1097 352 930 351 1205 350 1105 349 920 348 1379 347 1382 346 1133 345 1251 344 1445 343 1293 342 1381 341 1320 340 985 339 1041 338 919 337 912 239 816 238 1284 237 627 236 600 235 988 234 686 233 866 232 1087 231 920 230 596 229 782 228 1044 227 1105 226 1116 225 1102 224 812 223 1345 222 905 221 928 220 1275 219 948 218 791 217 1297 216 655 215 1024 214 1187 213 796 212 1190 211 771 210 897 209 1058 208 1361 207 778 206 808 205 1251 204 876 203 467 202 1330 201 840 200 997 199 959 198 1371 197 888 196 1177 195 1159 194 902 193 1289 192 1148 191 713 190 926 189 786 188 1006 187 811 186 849 185 1458 184 643 183 1286 182 1066 181 1078 180 957 179 1114 178 1122 177 934 176 1028 175 740 174 915 173 1006 172 1244 171 669 170 828 169 1475 168 784 167 1082 166 1297 165 1292 164 1139 163 1037 162 1135 161 1169 160 1068 159 974 158 863 157 1002 156 1324 155 868 154 510 153 1306 152 656 151 984 150 1326 149 1255 148 957 147 1130 146 1029 145 1160 144 1015 143 1119 142 803 141 913 140 739 139 1315 138 811 137 733 136 1350 135 1011 134 1060 133 922 132 1167 131 1273 130 1353 129 1060 128 913 127 963 126 1197 125 921 124 562 123 907 122 1101 121 813 120 1114 119 879 118 1140 117 1252 116 1085 115 1043 114 1150 113 1076 112 917 111 1102 110 777 109 846 108 841 107 1232 106 974 105 1008 104 904 103 613 102 1185 101 1202 100 1084 99 1290 98 864 97 1154 96 968 95 905 94 1106 93 1082 92 1055 91 1002 90 980 89 996 88 909 87 1186 86 730 85 1045 84 1015 83 1143 82 1275 81 1133 80 1176 79 432 78 1136 77 1084 76 1209 75 981 74 985 73 842 72 871 71 1056 70 1224 69 1068 68 1125 67 1026 66 1008 65 1098 64 1089 63 1289 62 794 61 1174 60 1075 59 1558 58 1257 57 1023 56 1146 55 914 54 952 53 1100 52 1172 51 1066 50 1307 49 850 48 1062 47 1212 46 1371 45 868 44 1080 43 1265 42 1288 41 1192 40 1524 39 1044 38 1034 37 1085 36 1143 35 1266 34 1013 33 1053 32 1066 31 1013 30 814 29 1049 28 1332 27 1263 26 917 25 1340 24 1140 23 1279 22 1004 21 1002 20 1133 19 1019 18 746 17 1025 16 1070 15 952 14 1038 13 898 12 663 11 1085 10 1162 9 1131 8 1146 7 1090 6 1277 5 942 4 1125 3 1023 2 554 1 1005

In [ ]:
%sql
DESCRIBE FORMATTED f1_processed.lap_times;

col_name data_type comment driver_id int null lap int null position int null time string null milliseconds int null ingestion_date timestamp null data_source string null file_date string null race_id int null # Partition Information # col_name data_type comment race_id int null # Detailed Table Information Catalog spark_catalog Database f1_processed Table lap_times Owner root Created Time Thu Jun 15 17:29:20 UTC 2023 Last Access UNKNOWN Created By Spark 3.3.2 Type MANAGED Provider parquet Location dbfs:/mnt/formula1dl/processed/lap_times Serde Library org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe InputFormat org.apache.hadoop.mapred.SequenceFileInputFormat OutputFormat org.apache.hadoop.hive.ql.io.HiveSequenceFileOutputFormat Partition Provider Catalog

<center><img src="https://i.postimg.cc/Dfp89bw7/db118.png"></center>

In [ ]:
dbutils.notebook.exit("Success")

Success